### DISCUSSION TITLE
dotabod integration

### DISCUSSION THREAD

**Discussion by Geczy** [AveYo/ShowMMR#6] [Ideas]
hey, i run dotabod.com 

any ideas how we can make this log mmr at scale, with thousands of users? im trying to find a nice streamlined process, but if one doesn't exist, i'm okay with accepting that

for example, does this app have to be updated with every steam / dota update for it to work? 

i could build a watch script that auto-updates itself on the computer when the vpk files change, which is one workaround for that issue

my concern is streamers (like gorgc and others) are gonna be like 'wtf' when they have to run a random .bat file (my installer.bat that i'd write for this repo). the installer would ideally take care of all the steps automatically, like adding launch options, saving files, building from vpk, etc. then it'd have to somehow send the mmr events to my dotabod backend server through an http request. is that possible using just this script or will i have to use windows terminal? like a backend daemon, or spawning a shell from your script, etc 

let me know if you're interested in something like that or have some answers

**Comment by AveYo** [comment]
Hi.
So you want to track the mmr progress for some users, correct me if I'm wrong.

One way is to run the **external** ShowMMR tool for each, it makes steam api calls via SteamKit2.
Credentials are cached locally for convenience, any subsequent runs only need the account name
Every evening could call `ShowMMR.exe thesteamuser anytext 10` to retrieve last 10 ranked matches in a file (output format can be easily adapted), no user input required. No need for the mod or dota at all.
But it's steam api calls with steam logins, and most people would not allow access to someone else.

I would instead ignore the optional tool meant to retrieve mmr numbers from before installing the dota client-side mod, and only passively monitor:
`dota 2 beta/game/dota/cfg/user_keys_12345678_slot3.vcfg`
for size changes (it gets refreshed often by the game)
where `12345678` being the steam_id for the user
After every ranked game, it should update on itself with the latest result.
No glaring account security issues, no api calls disruption / timeouts / 2fa issues.

The vcfg file is a combination of vdf and json (out of necessity), 
but should be trivial to extract and join data from `JOY1 = "data"` to `JOY32 = "data"`
then json.decode it, order it by keys (match timestamp) etc. 
the 1st entry in JOY1 should be the last mmr value, in the following example, 10322.
``` vcfg
"config"
{
	"bindings"
	{
		"JOY1"		"1681834589:[10322,-17],1681760961:[10339,60],1681730543:[10279,-20]"
	}
	"matches"
	{
		7123456786 { date 1681660525 	 mmr  10299 	 outcome    50 }
		7123456785 { date 1681634987 	 mmr  10249 	 outcome    29 }
		7123456784 { date 1681630569 	 mmr  10220 	 outcome    10 }
		7123456783 { date 1681560541 	 mmr  10210 	 outcome   -30 }
		7123456782 { date 1681460923 	 mmr  10240 	 outcome   -30 }
		7123456781 { date 1681364585 	 mmr  10270 	 outcome    30 }
		7123456780 { date 1681364567 	 mmr  10240 	 outcome    30 }
		7123456779 { date 1681360949 	 mmr  10210 	 outcome   -30 }
		7123456778 { date 1681360521 	 mmr  10240 	 outcome    30 }
		7123456777 { date 1681330583 	 mmr  10210 	 outcome    30 }
		7123456776 { date 1681264965 	 mmr  10180 	 outcome   -30 }
		7123456775 { date 1681234547 	 mmr  10210 	 outcome    30 }
		7123456774 { date 1681230529 	 mmr  10180 	 outcome    30 }
		7123456773 { date 1681134981 	 mmr  10150 	 outcome   -30 }
		7123456772 { date 1681060563 	 mmr  10180 	 outcome   -30 }
		7123456771 { date 1681060545 	 mmr  10210 	 outcome    30 }
		7123456770 { date 1681034927 	 mmr  10180 	 outcome   -30 }
	}
}
``` 

The mod does not replace frequently updated files so it should be resilient to most game updates
It's rather set it and forget it.

Ask away!

**Comment by Geczy** [comment]
very nice, the last idea seems good. just to confirm, that vcfg file is only created & then auto updated if each user goes through the `To install the mod:` steps?

also, any chance steam credentials could be an oauth token or similar? so through any responses from the "login with steam" button they may click on dotabod.com ? re `ShowMMR.exe thesteamuser anytext 10`

**Comment by AveYo** [comment]
Yes.  

The token is saved locally in username.auth but that's a full login token, not a restricted _login with steam_ for web apps that strip away most access - using it for unintended purposes sounds like opening a can of worms to me.
I just adapted the code from SteamKit2 samples so they might know better, tho I suspect they would advise the same, the SteamKit2 library is a steam emulator meant for personal usage and even that with no guarantees.

**Comment by AveYo** [comment]
I might overthink it - your plugin or whatever could do the same for the tool generated file that the user would run locally. It would not be much hindrance to click it once to generate the file since like I said, credentials are cached.
Then do a http post request with the resulting file to your end point, or whatever.
But you can't turn it into an api call to directly share data to your website cutting the local middleman.
Does this make any sense?

**Comment by Geczy** [comment]
i think i know what to do now! im coming back to this brilliant solution, thanks for your help.

i'm going to create a script that generates their custom .exe for them which includes their dotabod token, and calls it like ShowMMR.exe thedotabodtoken 10

does 'thesteamuser' actually need to be a steam username? because what if they have multiple steam accounts on the same computer? can the exe differentiate data by steam id?

**Reply by Geczy** [reply]
basically if the output data has the steam id included that would probably help, even grouped by it somewhere here

"config"
{
	"bindings"
	{
		"JOY1"		"1681834589:[10322,-17],1681760961:[10339,60],1681730543:[10279,-20]"
	}
	"matches"
	{
		7123456786 { date 1681660525 	 mmr  10299 	 outcome    50 }
		7123456785 { date 1681634987 	 mmr  10249 	 outcome    29 }
		7123456784 { date 1681630569 	 mmr  10220 	 outcome    10 }
		7123456783 { date 1681560541 	 mmr  10210 	 outcome   -30 }

**Reply by Geczy** [reply]
would i have to monitor all *.cfg files in

`dota 2 beta/game/dota/cfg/*.vcfg` ? are they created automatically if i just call one .exe ? or do i have to call many

**Reply by AveYo** [reply]
there is a file per each account `user_keys_123456789_slot3.vcfg` where 123456789 is the steamid
monitoring the cfg folder is the easiest (can sort by date to get the last logged on acc)

and yes these are created from a single built exe, just call run.bat with different arguments `steam-user` `pass` `optional-count-of-matches` `optional-start-from-match-id`

the pass / steamguard etc will be cached and won't be required on subsequent runs

just fixed the ShowMMR clientside mod

**Reply by Geczy** [reply]
hmm not sure if someone like gorgc is going to save `dotabod.exe gorgc <password>` , is there no way the exe can find their steam credentials on the pc already cached somewhere? like when they login to the steam client, isn't that session usable? 

what about 2fa or email auth? are those codes not triggered when this usr/pw is passed in? 

f password is required to be passed in im not sure if it's plausible anymore 😨
